### PR TITLE
Refactoring

### PR DESCRIPTION
**Pull request by vjik**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ❌
| Breaks BC?    | ❌
| Fixed issues  | -

`Normalizer::parse` from factory moved to new `DefinitionParser` in container.

Associated with https://github.com/yiisoft/factory/pull/85

### CODE REVIEW

**Review by yiiliveext — APPROVED** [review]

File: src/DefinitionParser.php [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+
+use Yiisoft\Factory\Definition\ArrayDefinition;
+use Yiisoft\Factory\Exception\InvalidConfigException;
+
+use function in_array;
+use function is_array;
+use function is_string;
+
+/**
+ * @internal Split metadata and definition
```
Comment (samdark) [inline src/DefinitionParser.php:15]:
```suggestion
 * @internal Splits metadata and definition.
```

File: src/DefinitionParser.php [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+ *
+ * Support configuration:
```
Comment (samdark) [inline src/DefinitionParser.php:17]:
```suggestion
 * Supports the following configuration:
```

**Review by samdark — APPROVED** [review]

**Review by rustamwin — APPROVED** [review]

File: src/DefinitionNormalizer.php [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+final class DefinitionNormalizer
+{
+    /**
+     * @param mixed $definition
+     *
+     * @throws InvalidConfigException
+     */
+    public static function normalize($definition, string $id = null): DefinitionInterface
+    {
+        if (is_array($definition) && isset($definition[DefinitionParser::IS_PREPARED_ARRAY_DEFINITION_DATA])) {
```
[the remark this answers is not in the document]
Comment (vjik) [reply src/DefinitionNormalizer.php:25]: No, if you use `IS_PREPARED_ARRAY_DEFINITION_DATA` in configuration, then this throws exception in `DefinitionParse::prepareDefinitionFromArray`.

File: src/Container.php [hunk 4]
```diff
@@ -211,6 +224,43 @@ private function setDefaultDefinitions(): void
...
+        }
+        DefinitionValidator::validate($definition, $id);
+    }
+
+    /**
+     * @throws InvalidConfigException
+     */
+    private function validateMeta(array $meta): void
+    {
+        foreach ($meta as $key => $_value) {
```
Comment (yiiliveext) [inline src/Container.php:250]: Maybe it will be better to use array_diff()

File: src/Container.php [hunk 5]
```diff
@@ -209,6 +224,43 @@ private function setDefaultDefinitions(): void
...
     }
 
+    /**
+     * @param mixed $definition
+     *
+     * @throws InvalidConfigException
+     */
+    private function validateDefinition($definition, ?string $id = null): void
+    {
+        if (is_array($definition) && isset($definition[DefinitionParser::IS_PREPARED_ARRAY_DEFINITION_DATA])) {
```
Comment (yiiliveext) [inline src/Container.php:234]: Too many operations for prepared definitions.

File: src/DefinitionNormalizer.php (hunk 3 above)
Comment (yiiliveext) [inline src/DefinitionNormalizer.php:25]: Too many operations for prepared definitions.

**Comment by vjik** [comment]
Successfully tested in app and demo.

**Review by samdark — APPROVED** [review]

**Comment by samdark** [comment]
👍

**Pull request merged**